File: LeetCode - Algorithm/Graph/207.py
# 207. Course Schedule
# https://leetcode.com/problems/course-schedule/

from typing import List
from collections import defaultdict


class Solution:
    # time complexity: O(n ^ 2)
    # Time Limit Exceeded :(
    # It can be improved by caching and memoization. The process currently rechecks whether a certain course
    # can be taken, even though it was already verified in a previous route, which is very wasteful.
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        prerequisiteGraph = defaultdict(list)

        for targetCourse, preCourse in prerequisites:
            prerequisiteGraph[targetCourse].append(preCourse)

        for i in range(numCourses):
            if not self.isPossibleToTake(prerequisiteGraph, i, set()):
                return False

        return True

    # I initially didn't call visited.remove(courseNum). When the task is to check for the presence of a cycle,
    # which means verifying a specific path, 'add' and 'remove' should be called consecutively.
    def isPossibleToTake(
        self,
        prerequisiteGraph: defaultdict[int, List[int]],
        courseNum: int,
        visited: set[int],
    ):
        if courseNum in visited:
            return False

        visited.add(courseNum)

        for course in prerequisiteGraph[courseNum]:
            if not self.isPossibleToTake(prerequisiteGraph, course, visited):
                return False

        visited.remove(courseNum)

        return True

File: LeetCode - Algorithm/Graph/207.js
// 207. Course Schedule
// https://leetcode.com/problems/course-schedule/

/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */

// Time Limit Exceeded :(
var canFinish1 = function (numCourses, prerequisites) {
  const coursesMap = {};
  prerequisites.forEach(([course, prerequisite]) => {
    if (!coursesMap[course]) coursesMap[course] = [];

    coursesMap[course].push(prerequisite);
  });

  const isThereLoop = (course, visited) => {
    if (visited.includes(course)) return true;

    const targetCourses = coursesMap[course];
    if (!targetCourses) return false;

    visited.push(course);

    let result = false;

    for (let i = 0; i < targetCourses.length; i++) {
      result = result || isThereLoop(targetCourses[i], visited);
      if (result) return result;
    }

    visited.pop();

    return result;
  };

  for (let i = 0; i < numCourses; i++) {
    if (isThereLoop(i, [])) return false;
  }

  return true;
};

var canFinish2 = function (numCourses, prerequisites) {
  const coursesMap = {};
  prerequisites.forEach(([course, prerequisite]) => {
    if (!coursesMap[course]) coursesMap[course] = [];

    coursesMap[course].push(prerequisite);
  });

  const okaySet = new Set();

  const isThereLoop = (course, visited) => {
    if (visited.includes(course)) return true;
    if (okaySet.has(course)) return false;

    const targetCourses = coursesMap[course];
    if (!targetCourses) return false;

    visited.push(course);

    let result = false;

    for (let i = 0; i < targetCourses.length; i++) {
      result = result || isThereLoop(targetCourses[i], visited);
      if (result) return result;
    }

    visited.pop();

    okaySet.add(course);

    return result;
  };

  for (let i = 0; i < numCourses; i++) {
    if (isThereLoop(i, [])) return false;
  }

  return true;
};

`
0 -> [1]
1 -> [0]

0 -> [1, 2]
1 -> [2, 3]
2 -> [3]

0 -> [5]
1 -> [0, 7]
2 -> [6]
6 -> [4]
7 -> [0]
`;
